<a href="https://colab.research.google.com/github/PPosadaQ/Analitica-de-Negocios---EAFIT/blob/main/Reto_2_%C3%81rboles_de_Decisi%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Árboles de Decisión**
0. Cargar las librerías de referencia para el modelo

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt #Graficador
from sklearn.tree import DecisionTreeClassifier


1. Se procede con la carga de los datos del archivo Travel_Plan_Score

In [ ]:
nxl= '/content/2. BD1_Travel_Data_int.xlsx'
XDB = pd.read_excel(nxl, sheet_name = 0)
XDB = XDB.dropna() #Elimina celdas vacías o sin datos
XDB.head(100)

College  CreditCard  FoodSpend  Income  TravelSpend
0         1           1    5472.43   49150       827.40
1         0           1    9130.73   47806       863.55
2         0           1    4450.67   46050      1180.05
3         0           1    5020.72   42600       755.70
4         0           1    3408.11   52092       851.40
..      ...         ...        ...     ...          ...
95        0           0    2019.98   65982      4338.75
96        0           0    7046.11   53986      1851.60
97        0           0    5180.13   53986      1646.70
98        0           0    3266.44   63962      1641.60
99        1           1    2082.23   49004      1564.95

[100 rows x 5 columns]

In [ ]:
#Se procede con la selección de las variables de trabajo
XD=np.array(XDB[['College','FoodSpend','Income','TravelSpend']]) #Datos de Entrada
print(XD)
yd=np.array(XDB[['CreditCard']]) #Este es el dato que quiero que el modelo pronostique
print(yd)

[[1.00000e+00 5.47243e+03 4.91500e+04 8.27400e+02]
 [0.00000e+00 9.13073e+03 4.78060e+04 8.63550e+02]
 [0.00000e+00 4.45067e+03 4.60500e+04 1.18005e+03]
 ...
 [0.00000e+00 3.58586e+03 4.22500e+04 2.25195e+03]
 [1.00000e+00 1.83395e+03 6.02480e+04 2.00025e+03]
 [0.00000e+00 6.20597e+03 5.85200e+04 3.66750e+03]]
[[1]
 [1]
 [1]
 [1]
 [1]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [1]
 [1]
 [1]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [0]
 [1]
 [1]
 [0]
 [1]
 [0]
 [1]
 [1]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [0]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0

2. Creación del modelo - Decision Tree

In [ ]:
mar=DecisionTreeClassifier(criterion='gini', max_depth = 4) #4 Niveles o variables de entrada
mar.fit(XD,yd)
ydp=mar.predict(XD)

from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)

#Evaluamos el modelo
VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

Ex= (VN+VP)/(VN+VP+FN+FP) #Que porcentaje de los datos pronosticó correctamente (Es VP y VN)
print("La exactitud alcanzada por el modelo es:",Ex)

Se= VP/(FN+VP) #Como le fue con los positivos (Es FN y VP) Total Aprobados
print("La sensibilidad alcanzada por el modelo fue:",Se)

Es= VN/(VN+FP) #Como le fue con los negativos (Es VN y FP) Total Prenegados
print("La especificidad alcanzada por el modelo es:",Es)

[[334   9]
 [ 42 115]]
La exactitud alcanzada por el modelo es: 0.898
La sensibilidad alcanzada por el modelo fue: 0.732484076433121
La especificidad alcanzada por el modelo es: 0.9737609329446064


3. ¿Y Donde está el Árbol?

In [ ]:
#Estas son las librerías para graficar el árbol
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data

VS = ['College','FoodSpend','Income','TravelSpend'] #Son los títulos para los nodos
dot_graph=export_graphviz(mar,feature_names=VS)
graph=graph_from_dot_data(dot_graph)
graph.write_png('arbol.png')

True

4. Se procede con el pronóstico de los nuevos solicitantes

In [ ]:
XDB2=pd.read_excel(nxl, sheet_name=1)
XDB2 = XDB2.dropna() #Elimina las celdas vacías o sin datos
XDB2.head(100)

College  FoodSpend  Income  TravelSpend
0        0    2892.90   65982       2151.6
1        1    6017.66   53986       2002.5

In [ ]:
XD2= np.array(XDB2[['College','FoodSpend','Income','TravelSpend']]) #Datos de entrada
print(XD2)

#Pronosticamos la PreAprobación
ydp2=mar.predict(XD2)
print(ydp2)
print("El número de PreAprobados es:" ,np.sum(ydp2))
print("El numero total de Datos es:" , len(ydp2))
print("El numero de PreNegados es:" , len(ydp2)-np.sum(ydp2))

[[0.00000e+00 2.89290e+03 6.59820e+04 2.15160e+03]
 [1.00000e+00 6.01766e+03 5.39860e+04 2.00250e+03]]
[0 0]
El número de PreAprobados es: 0
El numero total de Datos es: 2
El numero de PreNegados es: 2


**Análisis de Resultados**

De acuerdo con la base de datos suministrada y el nodo del árbol de decisión se puede observar un total de 500 datos, divididos en 343 PreAprobados y un total de 157 PreNegados, esto representa el 68,6% y el 31,4% respectivamente. De acuerdo con el pronóstico realizado por el modelo, este alcanzó una exactitud cercana al 89%, se puede considerar que es un modelo que tiene una tasa de exactitud ALTA. Por otro lado, se obtuvo una sensibilidad cercana al 73%, indicandonos que cerca del 27% de los datos que se deberían clasififcar como positivos se están clasificando erroneamente, clasificandolos como falsos negativos y por último se obtuvo una especificidad cercana al 94%

Se destacan los nodos puros, los cuales fueron 3 y son:

- "[PreNeg:108,PreApr:0]". Se debe cumplir la regla de:
 TravelSpend > 1567.875 AND Income > 59098.0

- "[PreNeg:0,PreApr:47]". Se debe cumplir con la regla de:
TravelSpend <= 1188.0 AND FoodSpend > 4990.73

- "[PreNeg:1,PreApr:0]" . Se debe cumplir conla regla de:
TravelSpend <= 1188.0 AND FoodSpend <= 4990.73 AND FoodSpend > 4972.55

Con respecto al pronóstico de los datos, se puede observar un total de 2536 individuos sin PreAprobación, de los cuales un total de 1260 pueden ser PreAprobados y un total de 1276 que pueden ser PreNegados.

**Respuestas del documento de WORD**

1. Las variables de entrada que conforman el modelo son: College, FoodSpend, Income y TravelSpend.
La variable de salida es CreditCard

2. La exactitud alcanzada por el modelo es: 0.898
La sensibilidad alcanzada por el modelo fue: 0.732484076433121
La especificidad alcanzada por el modelo es: 0.9737609329446064

3. Las características del primer cliente pertenecen al siguiente nodo puro "[PreNeg:108,PreApr:0]". Con estas características, se tiene la probabilidad del 100% de que el resultado sea PreNegado

Las características del segundo cliente pertenecen al siguiente nodo NO puro
"[PreNeg:107,PreApr:29]". Con estas características, se tiene la probabilidad del 78,68% de que el resultado sea PreNegado y la probabilidad del 21,32% de que sea PreAprobado.

4. Regla del negocio del PRIMER cliente:
TravelSpend > 1567.875 AND Income > 59098.0

Regla del negocio del SEGUNDO cliente:
TravelSpend > 1567.875 AND Income <= 59098.0 AND
FoodSpend > 2711.895 AND TRavel Spend <= 3008.25


5. Se destacan los nodos puros, los cuales fueron 3 y son:

"[PreNeg:108,PreApr:0]". Se debe cumplir la regla de: TravelSpend > 1567.875 AND Income > 59098.0

"[PreNeg:0,PreApr:47]". Se debe cumplir con la regla de: TravelSpend <= 1188.0 AND FoodSpend > 4990.73

"[PreNeg:1,PreApr:0]" . Se debe cumplir conla regla de: TravelSpend <= 1188.0 AND FoodSpend <= 4990.73 AND FoodSpend > 4972.55